core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -0,0 +1,202 @@
+          countDownLatch.countDown()
+        }}, ctx)
+      case SetDataRequest(path, data, version, ctx) => sessionContext.get().setData(path, data, version, new StatCallback {
+        override def processResult(rc: Int, path: String, ctx: Any, stat: Stat) = {
+          responseQueue.add(SetDataResponse(rc, path, ctx, stat))
+          countDownLatch.countDown()
+        }}, ctx)
+      case GetACLRequest(path, ctx) => sessionContext.get().getACL(path, null, new ACLCallback {
+        override def processResult(rc: Int, path: String, ctx: Any, acl: java.util.List[ACL], stat: Stat): Unit = {
+          responseQueue.add(GetACLResponse(rc, path, ctx, Option(acl).map(_.asScala).orNull, stat))
+          countDownLatch.countDown()
+        }}, ctx)
+      case SetACLRequest(path, acl, version, ctx) => sessionContext.get().setACL(path, acl.asJava, version, new StatCallback {
+        override def processResult(rc: Int, path: String, ctx: Any, stat: Stat) = {
+          responseQueue.add(SetACLResponse(rc, path, ctx, stat))
+          countDownLatch.countDown()
+        }}, ctx)
+      case GetChildrenRequest(path, ctx) => sessionContext.get().getChildren(path, false, new Children2Callback {
+        override def processResult(rc: Int, path: String, ctx: Any, children: java.util.List[String], stat: Stat) = {
+          responseQueue.add(GetChildrenResponse(rc, path, ctx, Option(children).map(_.asScala).orNull, stat))
+          countDownLatch.countDown()
+        }}, ctx)
+    }
+    countDownLatch.await()
+    responseQueue.asScala.toSeq
+  }
+
+  def getState: States = sessionContext.get().getState
+
+  def waitUntilConnectedOrExpired: Boolean = inLock(isConnectedOrExpiredLock) {
+    var state = sessionContext.get().getState
+    while (!state.isConnected && state.isAlive) {
+      isConnectedOrExpiredCondition.await()
+      state = sessionContext.get().getState
+    }
+    state.isConnected
+  }
+
+  def registerStateChangeHandler(stateChangeHandler: StateChangeHandler): Unit = {
+    stateChangeHandlerOpt = Option(stateChangeHandler)
+  }
+
+  def unregisterStateChangeHandler(): Unit = {
+    stateChangeHandlerOpt = None
+  }
+
+  def registerZNodeChangeHandler(zNodeChangeHandler: ZNodeChangeHandler): Unit = {
+    zNodeChangeHandlers.put(zNodeChangeHandler.path, zNodeChangeHandler)
+    sessionContext.get().exists(zNodeChangeHandler.path, true)
+  }
+
+  def unregisterZNodeChangeHandler(path: String): Unit = {
+    zNodeChangeHandlers.remove(path)
+  }
+
+  def registerZNodeChildChangeHandler(zNodeChildChangeHandler: ZNodeChildChangeHandler): Unit = {
+    zNodeChildChangeHandlers.put(zNodeChildChangeHandler.path, zNodeChildChangeHandler)
+    sessionContext.get().getChildren(zNodeChildChangeHandler.path, true)
+  }
+
+  def unregisterZNodeChildChangeHandler(path: String): Unit = {
+    zNodeChildChangeHandlers.remove(path)
+  }
+
+  def initialize(): Unit = inLock(initializationLock) {
+    if (!zooKeeper.getState.isAlive) {
```

Comment:
Hmm, but the check for zooKeeper.getState.isAlive is not going to be reliable since the state may not be alive immediately on a newly created session. Perhaps we can maintain a KeeperState that's updated through the watcher. Then, we can only create a new session if that state is in expiration state. After creating a new Zookeeper object, we can first set the state to Disconnected.